src/compiler/checker/lines_16000_18000.rs

```diff
@@ -132,6 +140,195 @@ impl TypeChecker {
+                let targets = &mapper.targets;
+                for (i, source) in sources.iter().enumerate() {
+                    if Rc::ptr_eq(&type_, source) {
+                        return if let Some(targets) = targets {
+                            targets[i].clone()
+                        } else {
+                            self.any_type()
+                        };
+                    }
+                }
+                type_
+            }
+            TypeMapper::Function(mapper) => (mapper.func)(self, type_),
+            TypeMapper::Composite(composite_or_merged_mapper)
+            | TypeMapper::Merged(composite_or_merged_mapper) => {
+                let t1 = self.get_mapped_type(type_.clone(), &composite_or_merged_mapper.mapper1);
+                if !Rc::ptr_eq(&t1, &type_) && matches!(mapper, TypeMapper::Composite(_)) {
+                    self.instantiate_type(Some(t1), Some(&composite_or_merged_mapper.mapper2))
+                        .unwrap()
+                } else {
+                    self.get_mapped_type(t1, &composite_or_merged_mapper.mapper2)
+                }
+            }
+        }
+    }
+
+    pub(super) fn make_unary_type_mapper(&self, source: Rc<Type>, target: Rc<Type>) -> TypeMapper {
+        TypeMapper::new_simple(source, target)
+    }
+
+    pub(super) fn make_array_type_mapper(
+        &self,
+        sources: Vec<Rc<Type /*TypeParameter*/>>,
+        targets: Option<Vec<Rc<Type>>>,
+    ) -> TypeMapper {
+        TypeMapper::new_array(sources, targets)
+    }
+
+    pub(super) fn make_composite_type_mapper(
+        &self,
+        mapper1: TypeMapper,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        TypeMapper::new_composite(mapper1, mapper2)
+    }
+
+    pub(super) fn make_merged_type_mapper(
+        &self,
+        mapper1: TypeMapper,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        TypeMapper::new_merged(mapper1, mapper2)
+    }
+
+    pub(super) fn combine_type_mappers(
+        &self,
+        mapper1: Option<TypeMapper>,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        if let Some(mapper1) = mapper1 {
+            self.make_composite_type_mapper(mapper1, mapper2)
+        } else {
+            mapper2
+        }
+    }
+
+    pub(super) fn merge_type_mappers(
+        &self,
+        mapper1: Option<TypeMapper>,
+        mapper2: TypeMapper,
+    ) -> TypeMapper {
+        if let Some(mapper1) = mapper1 {
+            self.make_merged_type_mapper(mapper1, mapper2)
+        } else {
+            mapper2
+        }
+    }
+
+    pub(super) fn instantiate_symbol(
+        &self,
+        mut symbol: Rc<Symbol>,
+        mapper: &TypeMapper,
+    ) -> Rc<Symbol> {
+        let links = self.get_symbol_links(&symbol);
+        let links = links.borrow();
+        if let Some(type_) = links.type_.as_ref() {
+            if !self.could_contain_type_variables(type_.clone()) {
+                return symbol;
+            }
+        }
```

Comment:
I don't know that it's an improvement, but looking for built-in ways to accomplish this same thing got me here:

```rs
let links = links.borrow();
if links
    .type_
    .as_ref()
    .filter(|type_| !self.could_contain_type_variables((*type_).clone()))
    .is_some()
{
    return symbol;
}
```

Generally I find if-statements about `Option`s/`Result`s to be kind of a smell, because there are so many convenience functions built in.

Reply:
Ah ya that's pretty nice. The thing that would read closest to the Typescript version would be:
```
if let Some(type_) = links.type_.as_ref() && !self.could_contain_type_variables(type_.clone()) {
  ...
}
```
but I think having "hybrid" `if let`/"normal" `if` conditions like that is only supported on nightly or something?

Reply:
Yeah I think so